### DISCUSSION TITLE
How do I modify my keybinding to place a full length (Ruler at 85) double comment block irrespective of current indentation?

### DISCUSSION THREAD

**Discussion by futuremotiondev** [ArturoDent/find-and-transform#28] [Q&A] [answered]
I am using this to create a TypeScript comment block:

```
{
    "key": "alt+c",
    "command": "findInCurrentFile",
    "when"    :  "editorTextFocus && !editorReadonly && editorLangId == typescript",
    "args": {
        "description": "Insert big full length comment block",
        "preCommands": "cursorHomeSelect",
        "replace": [
            "$${",
              "let str1 = '${selectedText} '.padEnd(85, '/');",
              "let str2 = '\n'.padEnd(86, '/');",
              "return str1 + str2",
            "}$$",
        ],
        "restrictFind": "selections",
        "postCommands": "cancelSelection", // to remove the resulting selection
    }
}
```

This results in:

```
// TEST  ////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////
```

But the block is misaligned if I trigger the shortcut in a further indented scope. See this video:

![FindAndTransformCommentBlocks](https://github.com/ArturoDent/find-and-transform/assets/33441569/3bdc95a8-7e34-4a6a-a331-ea45af5db701)

How can I modify this keybinding to make it work irrespective of current indentation?

**Comment by ArturoDent** [comment]
There are two ways to account for that leading indentation: (1) capture it or (2) do a format of the selection.

Try this keybinding for option (2):

```jsonc
{
 "key": "alt+c",
 "command": "findInCurrentFile",
 "when"    :  "editorTextFocus && !editorReadonly && editorLangId == typescript",
 "args": {
     "description": "Insert big full length comment block",
     "preCommands": "cursorHomeSelect",
     "replace": [
         "$${",
           "let str1 = '${selectedText} '.padEnd(85, '/');",
           "let str2 = '\n'.padEnd(86, '/');",
           "return str1 + str2",
         "}$$",
     ],
     "restrictFind": "selections",
     "postCommands": [
      "editor.action.formatSelection",           // add this
       "cancelSelection"
     ]
 }
}
```

Let me know if that works for you.  Option (1) is a little more involved but not too much.  You would capture the leading whitespace on the first line (where you typed `// TEST`) and add it to the start of the second line.

**Reply by futuremotiondev** [reply]
Great! Thank you for that. I will give it a try tomorrow and report back.

**Comment by ArturoDent** [comment, marked as answer]
Here is a more generic way to add the custom comments you want to do.  The `formatSelection` works fine but you couldn't actually indent those comments any arbitrary amount as the re-formatitng will remove extra leading whitespace if you try to indent too much beyond where a formatter thinks you should.

This keybinding does allow you to indent anywhere and then trigger your keybinding:

```jsonc
{
  "key": "alt+c",
  "command": "findInCurrentFile",
  // below when clause just removed for testing
  //  "when"    :  "editorTextFocus && !editorReadonly && editorLangId == typescript",
  "args": {
    "description": "Insert big full length comment block",
    "preCommands": [
      "cursorHomeSelect",
      "cursorHomeSelect"          // 2 cursorHomeSelect's will select all the line back to the 0 column
    ],
    "replace": [
      "$${",
        "let str1 = '${selectedText} '.padEnd(85, '/');",

        "const regex = /\\s*/g;",
        "const leadingSpaces = `${selectedText}`.match(regex)[0];",    // this gets the leading whiteSpace

        "let str2 = `\\n${leadingSpaces}`.padEnd(86, '/');",

        "return str1 + str2",
      "}$$",
    ],
    "restrictFind": "selections",
    "postCommands": "cancelSelection"
  }
}
```

![customCommentsIndent](https://github.com/ArturoDent/find-and-transform/assets/31813701/6ec0309f-6134-4ee7-9554-eb7e120c7342)

**Comment by futuremotiondev** [comment]
Hey again, I'm back! 😁

So, the code above is no longer working for me. I think there were some breaking changes in an update to the plugin.

This is the exact command that's defined in my `keybindings.json` file:

```json
{
    "key": "alt+c",
    "command": "findInCurrentFile",
    "when" :  "editorTextFocus && !editorReadonly && editorLangId == jsonc",
    "args" : {
        "description": "Insert big full length comment block",
        "preCommands": [
            "cursorHomeSelect",
            "cursorHomeSelect" // 2 cursorHomeSlect's will select all the line back to the 0 column
        ],
        "replace": [
            "$${",
            "let str1 = '${selectedText} '.padEnd(85, '/');",
            "const regex = /\\s*/g;",
            "const leadingSpaces = `${selectedText}`.match(regex)[0];", // this gets the leading whiteSpace
            "let str2 = `\\n${leadingSpaces}`.padEnd(86, '/');",
            "return str1 + str2",
            "}$$",
        ],
        "restrictFind": "selections",
        "postCommands": "cancelSelection"
    }
},
```

Here's a screen recording of what's occurring:

https://github.com/user-attachments/assets/bf52ca9f-2155-455b-ad27-0e73c3783cad

I don't think the `const leadingSpaces` variable assignment is correctly capturing the leading spaces of the selection.

There is some other weirdness going on as well if you watch the video.

Is there any way to fix this command?

I would really appreciate any help at all on this - I absolutely love the add-on and want to continue using it.

Thanks so much.

**Comment by ArturoDent** [comment]
I see you are selecting the text (in different ways) first before triggering the keybinding.  I should have mentioned that the keybinding I showed was designed to be used by placing the cursor at the end of the text first.  In any case, try modifying to this:

```jsonc
      "preCommands": [
          "cursorEnd",              // add this
          "cursorHomeSelect",
          "cursorHomeSelect"
      ],
```

Now it should work whether you select some of the text or not.

--------------------------------

I created another extension, [Comment Blocks](https://marketplace.visualstudio.com/items?itemName=ArturoDent.comment-blocks), to make what you are trying to do easier.  With this keybinding:

```jsonc
{
  "key": "alt+b",
  "command": "comment-blocks.createBlock",
  "args": {
    // "keepIndentation": true,    // defualt
    // "selectCurrentLine": true,  // default
    "lineLength": 90,
    "justify": "left",
    "startText": "${LINE_COMMENT}",
    "subjects": [
      "${selectedText}",
      ""
    ],

    "padLines": "/",
    "gapLeft": 2,
    "gapRight": 2
  }
}
```

you could do this:

![commentBlocks5](https://github.com/user-attachments/assets/8893a140-44dc-4294-8e46-42362d9e1cb4)

**Reply by futuremotiondev** [reply]
I love you! Thank you! I am going to install Comment Blocks right away and try it out.

**Comment by futuremotiondev** [comment]
Ok, so I tried Comment Blocks, and it's great! The only problem is that it doesn't seem to be honoring the `"lineLength": 85` value I put in. Here's a screen-cast:

https://streamable.com/q8ojlf

Here's the same video embedded in this post:

https://github.com/user-attachments/assets/9a20f322-cf85-4940-b80f-db58b38dfadb

The vertical gray line in my editor is at length 85. I was under the impression the comment should extend to that line.

Is there something I need to change?

**Comment by futuremotiondev** [comment]
I was able to get the line length of the comment to work by increasing `lineLength` to `112`. But then I tried to add another comment block directly below and the line length was messed up, and the addon didn't add the closing comment character (In this case `*/` since I'm editing a CSS document). 

Here's another screen recording:


https://github.com/user-attachments/assets/d94b6ea0-87b9-4740-90d6-5ff512f78442

Let me know if I'm doing anything wrong.

**Comment by futuremotiondev** [comment]
I also tried the following code with Find And Transform:

```jsonc
{
        "key": "alt+c",
        "command": "findInCurrentFile",
        "when" :  "editorTextFocus && !editorReadonly && editorLangId == css",
        "args" : {
            "description": "Insert big full length comment block",
            "preCommands": [
                "cursorEnd",
                "cursorHomeSelect",
                "cursorHomeSelect" // 2 cursorHomeSlect's will select all the line back to the 0 column
            ],
            "replace": [
                "$${",
                "let str1 = '/* ${selectedText} '.padEnd(85, '/');",
                "const regex = /\\s*/g;",
                "const leadingSpaces = `${selectedText}`.match(regex)[0];", // this gets the leading whiteSpace
                "let str2 = `\\n${leadingSpaces}`.padEnd(84, '/');",
                "let str3 = '*/';",
                "return str1 + str2 + str3",
                "}$$",
            ],
            "restrictFind": "selections",
            "postCommands": "cancelSelection"
        }
    },
```

Screen Recording Again:

https://github.com/user-attachments/assets/4fb1cd19-4f36-4cfb-ba78-b3bde481f367

I don't know why, but I thought that this line:

`"let str1 = '/* ${selectedText} '.padEnd(85, '/');",`

Would correctly insert a `/*` in front of the selected text, and it is, but it's also inserting an additional five spaces or so before `${selectedText}` which is confusing me. `'/* ${selectedText} '` only has once space between the `/*` and `${selectedText}` so I have no idea why it's doing that.

**Comment by futuremotiondev** [comment]
Yet another post, this new setting seems to be working a lot better:

```jsonc
{
        "key": "alt+c",
        "command": "comment-blocks.createBlock",
        "when" :  "editorTextFocus && !editorReadonly && editorLangId == css",
        "args": {
            // "keepIndentation": true,    // defualt
            // "selectCurrentLine": true,  // default
            "lineLength": 112,
            "justify": "left",
            "startText": "${LINE_COMMENT}",
            "endText": "${LINE_COMMENT}",
            "subjects": [
                "${selectedText}",
                ""
            ],
            "padLines": "/",
            "gapLeft": 0,
            "gapRight": 1
        }
    },
```

But it's not foolproof. The first comment block is created correctly. But if I add another comment block directly below it, the addon doesn't seem to insert the last three `///` on the first line, and also isn't inserting the ending ` */` to close the comment. 

Screen Recording:

https://github.com/user-attachments/assets/deb458bc-a8b6-4aef-b01e-78d94582514c

You'll also notice, if I add a line of CSS below the first comment and then create a comment block below that line of CSS, it works. But chaining comments without CSS in-between does not work.

Sorry for all the posts, just want to get this working. I need this functionality in my life!

**Comment by futuremotiondev** [comment]
Also tried this to get a three line comment block instead of two:

```jsonc
{
        "key": "alt+c",
        "command": "comment-blocks.createBlock",
        "when" :  "editorTextFocus && !editorReadonly && editorLangId == css",
        "args": {
            // "keepIndentation": true,    // defualt
            // "selectCurrentLine": true,  // default
            "lineLength": 112,
            "justify": "left",
            "startText": "${LINE_COMMENT}",
            "endText": "${LINE_COMMENT}",
            "subjects": [
                "${selectedText}",
                "",
                ""
            ],
            "padLines": "/",
            "gapLeft": 0,
            "gapRight": 1
        }
    },
```

Screen Recording:

https://github.com/user-attachments/assets/a5bee3ec-23ed-4185-8fdd-a2e13d46fe14

**Comment by ArturoDent** [comment]
I think all the Comment Blocks issues you are seeing are because `css` files don't actually have **line** comments - only block comments!  Here is a part of vscode's own language configuration file for css and js files:

```jsonc
{"comments":{"blockComment":["/*","*/"]},  // for css files

{ "comments": { "lineComment": "//", "blockComment": [ "/*", "*/" ] }  // for js files for example
```
Yet there are toggle/remove/add line comment commands for css files in vscode!  But it just uses the blockComment characters.  This extension did not work around the issue of using `lineComment` when the language doesn't support them very well - causing the various issues you saw because it couldn't resolve `${LINE_COMMENT}` in css files (because it looks at the file's language configuration to see what comment characters are defined for any language).

With the next version, v0.5.2, if you use 

```jsonc
        "startText": "${LINE_COMMENT}",            // you will get ${BLOCK_COMMENT_START} 
        // "startText": "${BLOCK_COMMENT_START}",

        "endText": "${LINE_COMMENT}",             // you will get ${BLOCK_COMMENT_END}
        // "endText": "${BLOCK_COMMENT_END}",
```
But better is to simply use the commented versions above instead.  Try using `${BLOCK_COMMENT_START}` and `${BLOCK_COMMENT_END}` instead and see if you get the results you expect (`lineLength`, etc. should work better).  And then upgrade to v0.5.2 (once it is released) and you should be able to use those or `${LINE_COMMENT}` in css files.  I may try to remove the `${LINE_COMMENT}` completion suggestion in languages that don't have them but there is a lot of overhead to that.
